src/notifications/NotificationsManager.ts

```diff
@@ -161,26 +353,210 @@ class NotificationsManager {
+    {
+      nodeId,
+      data,
+      retries = 64,
+    }: {
+      nodeId: NodeId;
+      data: NotificationData;
+      retries?: number;
+    },
+    tran?: DBTransaction,
+  ): Promise<{
+    notificationId: NotificationId;
+    sendProm: Promise<void>;
+  }> {
+    const nodeIdEncoded = nodesUtils.encodeNodeId(nodeId);
+    const notificationId = this.outboxNotificationIdGenerator();
+    const notificationIdEncoded =
+      notificationsUtils.encodeNotificationId(notificationId);
+    const notificationDb: NotificationDB = {
       typ: 'notification',
       data: data,
       isRead: false,
       iss: nodesUtils.encodeNodeId(this.keyRing.getNodeId()),
-      sub: nodesUtils.encodeNodeId(nodeId),
+      sub: nodeIdEncoded,
     };
-    const signedNotification = await notificationsUtils.generateNotification(
-      notification,
-      this.keyRing.keyPair,
+    let pendingTask: Task | undefined;
+    const transactionCallback = async (tran: DBTransaction) => {
+      await tran.put(
+        [
+          ...this.notificationsManagerOutboxNotificationsDbPath,
+          idUtils.toBuffer(notificationId),
+        ],
+        notificationDb,
+      );
+      const counter = await tran.get<number>(
+        this.notificationsManagerOutboxNotificationsCounterDbPath,
+      );
+      await tran.put(
+        this.notificationsManagerOutboxNotificationsCounterDbPath,
+        counter != null ? counter + 1 : 1,
+      );
+      pendingTask = await this.taskManager.scheduleTask(
+        {
+          handlerId: this.sendNotificationHandlerId,
+          parameters: [
+            {
+              nodeIdEncoded,
+              notificationIdEncoded,
+              retries,
+              attempt: 0,
+            },
+          ],
+          path: [this.sendNotificationHandlerId],
+          lazy: false,
+        },
+        tran,
+      );
+      this.outboxNotificationIdEncodedToTaskIdMap.set(
+        notificationIdEncoded,
+        pendingTask.id,
+      );
+    };
+    if (tran == null) {
+      await this.db.withTransactionF(transactionCallback);
+    } else {
+      await transactionCallback(tran);
+    }
+    const sendProm = (async () => {
+      while (pendingTask != null) {
+        pendingTask = await pendingTask.promise();
+      }
+    })();
+    sendProm.catch(() => {});
+    return {
+      notificationId,
+      sendProm,
+    };
+  }
+
+  protected async getOutboxNotificationIds(
+    tran: DBTransaction,
+  ): Promise<Array<NotificationId>> {
+    const notificationIds: Array<NotificationId> = [];
+    const messageIterator = tran.iterator<NotificationDB>(
+      this.notificationsManagerOutboxNotificationsDbPath,
+      { valueAsBuffer: false },
+    );
+    for await (const [keyPath] of messageIterator) {
+      const key = keyPath[0] as Buffer;
+      const notificationId = IdInternal.fromBuffer<NotificationId>(key);
+      notificationIds.push(notificationId);
+    }
+    return notificationIds;
+  }
```

Comment:
This should be a generator. Generally we want to avoid collecting generator outputs into an array unless we have to.

Reply:
this is partially done, I'm still fixing the tests for this.

Reply:
this has been done